File: QytetetJava/src/modeloqytetet/Qytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import javafx.util.Pair;
import InterfazTextualQytetet.VistaTextualQytetet;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.TreeMap;
/**
 *
 * @author jose26398
 */
public class Qytetet {
    static public int MAX_JUGADORES = 4;
    static final int MAX_CARTAS = 10;
    static final int MAX_CASILLAS = 20;
    static final int PRECIO_LIBERTAD = 200;
    static final int SALDO_SALIDA = 1000;
    private Dado dado = Dado.getInstance();
    private Sorpresa cartaActual;
    private Jugador jugadorActual;
    private ArrayList<Sorpresa> mazo = new ArrayList();
    private ArrayList<Jugador> jugadores = new ArrayList();
    private Tablero tablero;

    
    private Qytetet(ArrayList<String> nombres) {
        inicializarJuego(nombres);
    }

    public ArrayList<Jugador> getJugadores() {
        return jugadores;
    }
    
    public boolean aplicarSorpresa(){
        boolean tienePropietario = false;
        
        if (cartaActual.getTipo() == TipoSorpresa.PAGARCOBRAR)
            jugadorActual.modificarSaldo(cartaActual.getValor());
        
        else if(cartaActual.getTipo() == TipoSorpresa.IRACASILLA){
            boolean esCarcel = tablero.esCasillaCarcel(cartaActual.getValor());
            
            if(esCarcel){
                encarcelarJugador();
            }
            else{
                Casilla nuevaCasilla = tablero.obtenerCasillaNumero(cartaActual.getValor());
                tienePropietario = jugadorActual.actualizarPosicion(nuevaCasilla);
            }
        }
        else if(cartaActual.getTipo()==TipoSorpresa.PORCASAHOTEL){
            jugadorActual.pagarCobrarPorCasaYHotel(cartaActual.getValor());
            
        }
        else if(cartaActual.getTipo()==TipoSorpresa.PORJUGADOR){
            for(Jugador jugador: jugadores){
                if(jugador != jugadorActual){
                    jugador.modificarSaldo(cartaActual.getValor());
                    jugadorActual.modificarSaldo(-cartaActual.getValor());
                }
            }
        }
        
        if(cartaActual.getTipo() == TipoSorpresa.SALIRCARCEL)
            jugadorActual.setCartaLibertad(cartaActual);
        
        else{
            mazo.add(cartaActual);
            cartaActual = null;
        }
        
        return tienePropietario;
    }
    
    public boolean cancelarHipoteca(Casilla casilla){
        boolean puedoDeshipotecar = false;
        if(casilla.soyEdificable()){
            boolean sePuedeDeshipotecar = casilla.estaHipotecada();
            
            if(sePuedeDeshipotecar){
                puedoDeshipotecar = jugadorActual.puedoPagarHipoteca(casilla);
                
                if(puedoDeshipotecar){
                    int cantidadPagar = casilla.cancelarHipoteca();
                    jugadorActual.modificarSaldo(-cantidadPagar);
                }
            }
        }
        return puedoDeshipotecar;
    }
    
    public boolean comprarTituloPropiedad(){
        return jugadorActual.comprarTitulo();
    }
    
    public boolean edificarCasa(Casilla casilla){
        boolean puedoEdificar = false;
        
        if(casilla.soyEdificable()){
            boolean sePuedeEdificar = casilla.sePuedeEdificarCasa();
            
            if(sePuedeEdificar){
                puedoEdificar = jugadorActual.puedoEdificarCasa(casilla);
                
                if(puedoEdificar){
                    int costeEdificarCasa = casilla.edificarCasa();
                    jugadorActual.modificarSaldo(-costeEdificarCasa);
                }
            }
        }
        return puedoEdificar;
    }
    
    public boolean edificarHotel(Casilla casilla){
        boolean puedoEdificar = false;
        
        if(casilla.soyEdificable()){
            boolean sePuedeEdificar = casilla.sePuedeEdificarHotel();
            
            if(sePuedeEdificar){
                puedoEdificar = jugadorActual.puedoEdificarHotel(casilla);
                
                if(puedoEdificar){
                    int costeEdificarHotel = casilla.edificarHotel();
                    jugadorActual.modificarSaldo(-costeEdificarHotel);
                }
            }
        }
        return puedoEdificar;
    }

    public Sorpresa getCartaActual() {
        return cartaActual;
    }

    public Jugador getJugadorActual() {
        return jugadorActual;
    }
    
    public boolean hipotecarPropiedad(Casilla casilla){
        boolean puedoHipotecar = false;
        if(casilla.soyEdificable()){
            boolean sePuedeHipotecar = !(casilla.estaHipotecada());
            
            if(sePuedeHipotecar){
                puedoHipotecar = jugadorActual.puedoHipotecar(casilla);
                
                if(puedoHipotecar){
                    int cantidadRecibida = casilla.hipotecar();
                    jugadorActual.modificarSaldo(cantidadRecibida);
                }
            }
        }
        return puedoHipotecar;
    }
    
    private void inicializarJuego(ArrayList<String> nombres){
        if ( nombres.size() >= 2 && nombres.size() <= MAX_JUGADORES ){
            inicializarSorpresas();
            inicializarTablero();
            inicializarJugadores(nombres);
            salidaJugadores();            
        }
    }
    
    public boolean intentarSalirCarcel(MetodoSalirCarcel metodo){
        boolean libre = false;
        if(metodo == MetodoSalirCarcel.TIRANDODADO){
            int valorDado = dado.tirar();
            libre = valorDado > 5;
        }
        else if (metodo == MetodoSalirCarcel.PAGANDOLIBERTAD){
            boolean tengoSaldo = jugadorActual.pagarLibertad(PRECIO_LIBERTAD);
            libre = tengoSaldo;
        }
        
        if(libre)
            jugadorActual.setEncarcelado(false);
        
        return libre;
    }
    
    public boolean jugar(){        
        int valorDado = dado.tirar();
        Casilla casillaPosicion = jugadorActual.getCasillaActual();
        Casilla nuevaCasilla = tablero.obtenerNuevaCasilla(casillaPosicion, valorDado);
        boolean tienePropietario = jugadorActual.actualizarPosicion(nuevaCasilla);
        
        if(!nuevaCasilla.soyEdificable()){
            if(nuevaCasilla.getTipo() == TipoCasilla.JUEZ)
                encarcelarJugador();
            else if (nuevaCasilla.getTipo() == TipoCasilla.SORPRESA){
                cartaActual = mazo.get(0);
                mazo.remove(cartaActual);
                VistaTextualQytetet.Mostrar("La carta sorpresa es: " + cartaActual.toString());
            }
        }
        return tienePropietario;
    }
    
    public TreeMap obtenerRanking(){
        TreeMap<Integer,String> ranking = new TreeMap<>();
        
        for(Jugador jugador: jugadores){
           ranking.put(jugador.obtenerCapital(), jugador.getNombre());
        }
        return ranking;
    }
    
    public ArrayList<Casilla> propiedadesHipotecadasJugador(boolean hipotecadas){
        ArrayList<TituloPropiedad> titulos = jugadorActual.obtenerPropiedadesHipotecadas(hipotecadas);
        ArrayList<Casilla> devolver = new ArrayList();
        for(int i = 0; i < titulos.size(); i++)
            devolver.add(titulos.get(i).getCasilla());
        return devolver;
    }
    
    public void siguienteJugador(){        
        jugadorActual = jugadores.get((jugadores.indexOf(jugadorActual)+1)%jugadores.size());
    }
    
    public boolean venderPropiedad(Casilla casilla){
        boolean puedoVender = false;
        if(casilla.soyEdificable()){
            puedoVender = jugadorActual.puedoVenderPropiedad(casilla);
            
            if(puedoVender){
                jugadorActual.venderPropiedad(casilla);
            }
        }
        return puedoVender;
    }
    
    private void encarcelarJugador(){
        if(!jugadorActual.tengoCartaLibertad()){
            Casilla casillaCarcel = tablero.getCarcel();
            jugadorActual.irACarcel(casillaCarcel);
        }
        else{
            Sorpresa carta = jugadorActual.devolverCartaLibertad();
            mazo.add(carta);
        }
        
    }
    
    private void inicializarSorpresas(){
        mazo.add(new Sorpresa ("¡Ve a la cárcel! Vaya directamente a la caŕcel sin"
        + " pasar por la casilla de salida y sin cobrar los 1.000€", 5, TipoSorpresa.IRACASILLA));

        mazo.add(new Sorpresa ("Un fan anónimo ha pagado tu fianza. Sales "
        + "de la cárcel.", 0, TipoSorpresa.SALIRCARCEL));
        
        mazo.add(new Sorpresa ("Avanza hasta la casilla de salida", 0,
        TipoSorpresa.IRACASILLA));
        
        mazo.add(new Sorpresa ("Vas de viaje a la Calle de Velázquez.", 2,
        TipoSorpresa.IRACASILLA));
        
        mazo.add(new Sorpresa ("Multa por embriaguez. Page 600€", -600,
        TipoSorpresa.PAGARCOBRAR));
        
        mazo.add(new Sorpresa ("Cobra una herencia de 1.000€", 1000,
        TipoSorpresa.PAGARCOBRAR));
        
        mazo.add(new Sorpresa ("Haga reparaciones en todas sus casas."
        + "Page por cada casa 200€", -300, TipoSorpresa.PORCASAHOTEL));
        
        mazo.add(new Sorpresa ("Ganas un premio por el diseño de tus casas."
        + "Recibes 100€ por cada casa.", 200, TipoSorpresa.PORCASAHOTEL));
        
        mazo.add(new Sorpresa ("Hoy es tu cumpleaños. Por cada jugador recibe 150€",
        150, TipoSorpresa.PORJUGADOR));
        
        mazo.add(new Sorpresa ("Invitas a una cena a todos los jugadores."
        + "Das 200€ a cada jugador.", -200, TipoSorpresa.PORJUGADOR));
        
        //Collections.shuffle(mazo);
    }
    
   
    private void inicializarTablero(){
        tablero = new Tablero();
    }
    
    
    private void inicializarJugadores(ArrayList<String> nombres){
        for (int i=0; i<nombres.size(); i++)
            jugadores.add(new Jugador(nombres.get(i), tablero.obtenerCasillaNumero(0)));  
    }
    
    
    public static Qytetet getInstance(ArrayList<String> nombres) {
        if (instance == null){
            instance = new Qytetet(nombres);
            return instance;
        }
        else
            return instance;
    }
    
    private static Qytetet instance = null;

    @Override
    public String toString() {
        return "Qytetet{ MAX_JUGADORES=" + MAX_JUGADORES + ", MAX_CARTAS=" + MAX_CARTAS + ", MAX_CASILLAS=" + MAX_CASILLAS + ", PRECIO_LIBERTAD=" + PRECIO_LIBERTAD + ", SALDO_SALIDA=" + SALDO_SALIDA + "\n\nnombres=" + jugadores + "\n\ncartaActual=" + cartaActual + ", jugadorActual="+ jugadorActual + "\n\nmazo=" + mazo + "\n\ntablero=" + tablero + " } \n";
    }
    
    private void salidaJugadores(){
        for(int i = 0; i < jugadores.size(); i++){
            jugadores.get(i).modificarSaldo(7500 - jugadores.get(i).getSaldo());
            jugadores.get(i).actualizarPosicion(tablero.obtenerCasillaNumero(0));
        }
        jugadorActual = jugadores.get((int)(Math.random() * jugadores.size()));
    }
}

File: QytetetRuby/lib/qytetet.rb
#encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require "singleton"

module ModeloQytetet
  
  class Qytetet

    include Singleton
   
    @@MAX_JUGADORES = 4
    @@MAX_CARTAS = 10
    @@MAX_CASILLAS = 20    
    @@PRECIO_LIBERTAD = 200
    @@SALDO_SALIDA = 1000
    @@nombres = [];

    
    private
    def initialize()
      @cartaActual
      @mazo = []
      @jugadorActual
      @tablero
      @jugadores = []
      @dado = Dado.instance
    
      inicializar_juego(@@nombres);
      
    end

    
    public
    attr_reader :cartaActual, :jugadorActual, :mazo, :jugadores

    
    def self.asignar_nombres(nombres)
      @@nombres = nombres
    end
    
    
    def aplicar_sorpresa()
      tiene_propietario = false;

      if @cartaActual.tipo == TipoSorpresa::PAGARCOBRAR
        @jugadorActual.modificar_saldo( @cartaActual.valor )
      
      elsif @cartaActual.tipo == TipoSorpresa::IRACASILLA
        es_carcel = @tablero.es_casilla_carcel(@cartaActual.valor)
        
        if es_carcel == true
          encarcelar_jugador
        
        else
          nuevaCasilla = @tablero.obtener_casilla_numero(@cartaActual.valor)
          tiene_propietario = @jugadorActual.actualizar_posicion(nuevaCasilla)
        end
        
      elsif @cartaActual.tipo == TipoSorpresa::PORCASAHOTEL
        @jugadorActual.pagar_cobrar_por_casa_y_hotel(@cartaActual.valor)
      
      elsif @cartaActual.tipo == TipoSorpresa::PORJUGADOR
        for jugador in @jugadores
          if jugador != @jugadorActual
            jugador.modificar_saldo(@cartaActual.valor)
            @jugadorActual.modificar_saldo(-(@cartaActual.valor))
          end
        end
      end
      
      if @cartaActual.tipo == TipoSorpresa::SALIRCARCEL
        @jugadorActual.cartaLibertad = @cartaActual
      else
        @mazo << @cartaActual 
      end
      
      return tiene_propietario
    end

    
    def self.get_saldo_salida
      @@SALDO_SALIDA
    end

    
    def cancelar_hipoteca(casilla)
      puedo_deshipotecar = false
      
      if casilla.soy_edificable()
        se_puede_deshipotecar = (casilla.esta_hipotecada())
        
        if se_puede_deshipotecar
          puedo_deshipotecar = @jugadorActual.puedo_pagar_hipoteca(casilla)
          
          if puedo_deshipotecar
            cantidad_pagar = casilla.cancelar_hipoteca()
            @jugadorActual.modificar_saldo(-cantidad_pagar)
          end
        end
      end
      
      return puedo_deshipotecar
    end


    def comprar_titulo_propiedad()
      return @jugadorActual.comprar_titulo()
    end


    def edificar_casa(casilla)
      puedo_edificar = false
      
      if casilla.soy_edificable() == true
        se_puede_edificar = casilla.se_puede_edificar_casa()
        
        if se_puede_edificar == true
          puedo_edificar = @jugadorActual.puedo_edificar_casa(casilla)
          
          if puedo_edificar == true
            coste_edificar_casa = casilla.edificar_casa()
            @jugadorActual.modificar_saldo(-coste_edificar_casa)
          end
        end
      end
      
      return puedo_edificar
    end


    def edificar_hotel(casilla)
      puedo_edificar = false
      
      if casilla.soy_edificable() == true
        se_puede_edificar = casilla.se_puede_edificar_hotel()
        
        if se_puede_edificar == true
          puedo_edificar = @jugadorActual.puedo_edificar_hotel(casilla)
          
          if puedo_edificar == true
            coste_edificar_hotel = casilla.edificar_hotel()
            @jugadorActual.modificar_saldo(-coste_edificar_hotel)
          end
        end
      end
      
      return puedo_edificar
    end


    def hipotecar_propiedad(casilla)
      puedo_hipotecar = false
      
      if casilla.soy_edificable()
        se_puede_hipotecar = !(casilla.esta_hipotecada())
        
        if se_puede_hipotecar
          puedo_hipotecar = @jugadorActual.puedo_hipotecar(casilla)
          
          if puedo_hipotecar
            cantidad_recibida = casilla.hipotecar()
            @jugadorActual.modificar_saldo(cantidad_recibida)
          end
        end
      end
      
      return puedo_hipotecar
    end


    def inicializar_juego(nombres)
      if nombres.size >= 2 and nombres.size <= @@MAX_JUGADORES
        inicializar_cartas_sorpresa
        inicializar_tablero
        inicializar_jugadores(nombres)
        salida_jugadores
      end
    end


    def intentar_salir_carcel(metodo)
      libre = false
      
      if metodo == MetodoSalirCarcel::TIRANDODADO
        valor_dado = @dado.tirar()
        libre = (valor_dado > 5)
         
      elsif metodo == MetodoSalirCarcel::PAGANDOLIBERTAD
        tengo_saldo = @jugadorActual.pagar_libertad(@@PRECIO_LIBERTAD)
        libre = tengo_saldo
      end
      
      if libre == true
        @jugadorActual.encarcelado = false
      end
      
      return libre
    end


    def jugar()
      valor_dado = @dado.tirar()
      gets.chomp
      
      casilla_posicion = @jugadorActual.casillaActual
      nueva_casilla = @tablero.obtener_nueva_casilla(casilla_posicion, valor_dado)
      tiene_propietario = @jugadorActual.actualizar_posicion(nueva_casilla)
      
      if not nueva_casilla.soy_edificable()
        
        if nueva_casilla.tipo == TipoCasilla::JUEZ
          encarcelar_jugador()
          print("\nEl JUEZ te ha mandado a la carcel")
        elsif nueva_casilla.tipo == TipoCasilla::SORPRESA
          @cartaActual = @mazo[0]
          @mazo.delete_at(0)
        end
        
      end
      
      return tiene_propietario
    end


    def obtener_ranking()
      ranking = {}
      jugadores = @jugadores.sort{|j1,j2| j2.obtener_capital - j1.obtener_capital}

      for jugador in jugadores
        capital = jugador.obtener_capital
        ranking[jugador.nombre] = capital
      end
      
      return ranking
    end


    def propiedades_hipotecadas_jugador(hipotecadas)
      devolver = []
      @jugadorActual.obtener_propiedades_hipotecadas(hipotecadas).each{
        |aniadir|
        devolver << aniadir.casilla
      }
      return devolver
    end


    def siguiente_jugador()
      @jugadorActual = @jugadores[(@jugadores.index(@jugadorActual )    + 1) % @jugadores.size()]
      return @jugadorActual
    end


    def vender_propiedad(casilla)
      puedo_vender = false
      
      if (casilla.soy_edificable())
        puedo_vender = @jugadorActual.puedo_vender_propiedad(casilla)
        
        if puedo_vender
          @jugadorActual.vender_propiedad(casilla)
        end
      end
      
      return puedo_vender
    end
    
    
    private
    def encarcelar_jugador()
      
      if not @jugadorActual.tengo_carta_libertad
        casilla_carcel = @tablero.carcel;
        @jugadorActual.ir_a_carcel(casilla_carcel)
        
      else
        carta = @jugadorActual.devolver_carta_libertad
        @mazo << carta
      end

    end


    def inicializar_cartas_sorpresa()
      @mazo << Sorpresa.new("Ve a la carcel! Vaya directamente a la carcel sin
      pasar por la casilla de salida y sin cobrar los 1.000 euros", 5, TipoSorpresa::IRACASILLA)

      @mazo << Sorpresa.new("Un fan anonimo ha pagado tu fianza. Sales
      de la cárcel.", 0, TipoSorpresa::SALIRCARCEL)

      @mazo << Sorpresa.new("Avanza hasta la casilla de salida", 0,
      TipoSorpresa::IRACASILLA)

      @mazo << Sorpresa.new("Vas de viaje a la Calle de Velázquez", 2, 
      TipoSorpresa::IRACASILLA)

      @mazo << Sorpresa.new("Multa por embriaguez. Page 600€", -600,
      TipoSorpresa::PAGARCOBRAR)

      @mazo << Sorpresa.new("Cobra una herencia de 1.000€", 1000,
      TipoSorpresa::PAGARCOBRAR)

      @mazo << Sorpresa.new("Haga reparaciones en todas sus casas.
      Page por cada casa 200€", -300, TipoSorpresa::PORCASAHOTEL)

      @mazo << Sorpresa.new("Ganas un premio por el diseño de tus casas.
      Recibes 100€ por cada casa.", 200, TipoSorpresa::PORCASAHOTEL)

      @mazo << Sorpresa.new("Hoy es tu cumpleaños. Por cada jugador recibe 150€",
      -150, TipoSorpresa::PORJUGADOR)

      @mazo << Sorpresa.new("Invitas a una cena a todos los jugadores.
      Das 200€ a cada jugador.", 200, TipoSorpresa::PORJUGADOR)

      @mazo.shuffle!
    end


    def inicializar_jugadores(nombres)
      for i in nombres
        @jugadores << Jugador.new(i, @tablero.obtener_casilla_numero(0))
      end
    end


    def inicializar_tablero()
      @tablero = Tablero.new()
    end


    def salida_jugadores()
      @jugadores.each{
        |jug|
        jug.modificar_saldo(7500 - jug.saldo)
        jug.actualizar_posicion(@tablero.obtener_casilla_numero(0))
      }
      @jugadorActual = @jugadores[rand(@jugadores.size())]
    end
    
    
    public
    def to_s()
      "Qytetet{ MAX_JUGADORES: #{@@MAX_JUGADORES}, MAX_CARTAS: #{@@MAX_CARTAS}, MAX_CASILLAS: #{@@MAX_CASILLAS},
      PRECIO_LIBERTAD: #{@@PRECIO_LIBERTAD}, SALDO_SALIDA: #{@@SALDO_SALIDA} \n
      jugadores #{@jugadores} \n cartaActual: #{@cartaActual}, jugadorActual: #{@jugadorActual} \n
      mazo: #{@mazo} \n tablero: #{@tablero} } \n"
    end

  end
  
end